sched/task/task_setup.c

```diff
@@ -170,6 +170,16 @@ static int nxtask_assign_pid(FAR struct tcb_s *tcb)
 
   for (i = 0; i < g_npidhash / 2; i++)
     {
+      if (g_pidhash[i] == NULL)
+        {
+          /* If the pid is not used, skip it.
+           * This may be triggered when a context switch occurs
+           * during mm_malloc and a thread is destroyed.
+           */
```

Comment:
If in this case, we need to iterate over all g_npidhash instead of rebuilding the old half.

Reply:
But if a new pid is assigned, the reorder here will skip the second half

Reply:
This is already in the critical section, no thread switch will occur causing pid to be assigned

Reply:
>This is already in the critical section, no thread switch will occur causing pid to be assigned

as you mentioned, **already in the critical section, no thread switch will occur causing pid to be assigned**
so why needs to check NULL here?

Reply:
Thread switching may occur during malloc, there are two situations that need to be handled:

1. When switching, a new thread is created and g_pidhash is also expanded. When returning, g_pidhash has changed and there may be empty elements in it. At this time, accessing g_pidhash will access an illegal pointer
![image](https://github.com/user-attachments/assets/e1118231-9dd1-42a4-84f3-2bbf03d217dc)

3. When switching, a thread exits, a member of g_pidhash is cleared, which can also lead to illegal pointer access in subsequent iterations.
![image](https://github.com/user-attachments/assets/6c53f7ef-d399-407e-8981-49335be886dd)

Since thread switching can only occur during the malloc/free process, the other parts are protected by the critical section and thread switching will not occur.